File: python/open_guardrail/guards/tone_professional.py
"""Check for professional tone."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_SLANG: list[re.Pattern[str]] = [
    re.compile(
        r"\b(lol|lmao|rofl|omg|brb|tbh|fyi|btw"
        r"|idk|ngl|smh|imo|imho|af|tho|gonna"
        r"|wanna|gotta|kinda|sorta)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\b(dude|bro|fam|yolo|lit|slay|vibe"
        r"|sus|cap|no\s+cap|bet|salty|flex)\b",
        re.IGNORECASE,
    ),
]

_ALL_CAPS = re.compile(r"\b[A-Z]{4,}\b")


class _ToneProfessional:
    def __init__(
        self,
        *,
        action: str = "warn",
        max_exclamations: int = 2,
    ) -> None:
        self.name = "tone-professional"
        self.action = action
        self._max_excl = max_exclamations

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        for pat in _SLANG:
            if pat.search(text):
                issues.append("informal_language")

        excl = text.count("!")
        if excl > self._max_excl:
            issues.append("excessive_exclamations")

        if _ALL_CAPS.search(text):
            issues.append("all_caps_shouting")

        unique = list(dict.fromkeys(issues))
        triggered = len(unique) > 0
        score = (
            min(len(unique) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="tone-professional",
            passed=not triggered,
            action=(
                self.action if triggered else "allow"
            ),
            score=score,
            message=(
                f"Unprofessional tone: "
                f"{', '.join(unique)}"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "issues": unique,
                    "exclamation_count": excl,
                }
                if triggered
                else None
            ),
        )


def tone_professional(
    *,
    action: str = "warn",
    max_exclamations: int = 2,
) -> _ToneProfessional:
    return _ToneProfessional(
        action=action,
        max_exclamations=max_exclamations,
    )

File: packages/guards/src/tone-professional.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ToneProfessionalOptions {
  action: 'block' | 'warn';
  maxExclamations?: number;
}

const SLANG_PATTERNS: RegExp[] = [
  /\b(lol|lmao|rofl|omg|brb|tbh|fyi|btw|idk|ngl|smh|imo|imho|af|tho|gonna|wanna|gotta|kinda|sorta)\b/i,
  /\b(dude|bro|fam|yolo|lit|slay|vibe|sus|cap|no\s+cap|bet|salty|flex)\b/i,
];

const ALL_CAPS_PATTERN = /\b[A-Z]{4,}\b/;

export function toneProfessional(options: ToneProfessionalOptions): Guard {
  const maxExclamations = options.maxExclamations ?? 2;

  return {
    name: 'tone-professional',
    version: '0.1.0',
    description: 'Checks for professional tone in responses',
    category: 'content',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      for (const pattern of SLANG_PATTERNS) {
        if (pattern.test(text)) issues.push('informal_language');
      }

      const exclamations = (text.match(/!/g) || []).length;
      if (exclamations > maxExclamations) issues.push('excessive_exclamations');

      if (ALL_CAPS_PATTERN.test(text)) issues.push('all_caps_shouting');

      const unique = [...new Set(issues)];
      const triggered = unique.length > 0;
      const score = triggered ? Math.min(unique.length / 3, 1.0) : 0;

      return {
        guardName: 'tone-professional',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        message: triggered ? `Unprofessional tone: ${unique.join(', ')}` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues: unique, exclamationCount: exclamations } : undefined,
      };
    },
  };
}
